tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -526,6 +526,21 @@ void promoteToSharedGreedy(
     scop.insertSyncsAroundCopies(bandNode);
   }
 }
+
+// Returns a union of all mapping filters to "MappingType" in "scop".
+template <typename MappingType>
+isl::union_set collectMappingsTo(const Scop& scop) {
+  auto root = scop.scheduleRoot();
+  auto domain = scop.domain();
+  auto mappingFilters = detail::ScheduleTree::collect(
+      root, detail::ScheduleTreeType::MappingFilter);
+  mappingFilters = functional::Filter(isMappingTo<MappingType>, mappingFilters);
```

Comment:
Ha!  Why don't you do this in `extractDomainToIds` instead of the explicit `hasIds` stuff?

Reply:
Because a mapping to "t0" only is still a mapping to threads, but want that function to throw in case the caller asked for the mapping to "t0" _and_ "t1".

Reply:
I don't think any such distinction should be made.  A mapping node should always map to all (relevant) ids of a type.  It would be better to have a single way to check for a mapping node
than to have several ways of doing it slightly differently.  Anyway, I'll just leave it like this for now.